### ISSUE TITLE
[Question] Addings attributes in the songs pickle

### ISSUE THREAD

**Issue by petko10** [quodlibet/quodlibet#3127]
I'm planning on making an extension, that stores alternative URIs for songs (YouTube, SoundCloud, etc.), I'll give some more detail at the end. I was going to make a separate database in an e.g. JSPF file, since I want to also integrate exporting XSPF/JSPF playlists, but I figured it would be easier to keep the data in the pickled quodlibet database (if that is advisable). I currently don't know how to access the database TBH, and don't know if that's possible from an extension. I saw that there is a query interface, but I also read about ID3 being a limiting factor on tags, so I'm not sure if I could store my info as tag or if that's at all a proper way to handle the situation.

TL;DR: What's the proper way to store additional metadata about the songs with an extension?

About the planned setup (if anyone would find it interesting): I've grown frustrated that I can't have portable playlists and add online resources to them. So the plan is to make one or several quodlibet extensions, as well as a chrome extension and android app. The chrome extension would be for adding songs from various web services to a selected playlist (one at a time initially, automatic playlist sync with the streaming services will probably remain a dream). The chrome extension will just send the URLs to the quodlibet extension (probably via websocket). There the song will be cached locally via youtube-dl and there will be a wizard to fix the metadata with the initial import (optionally). Since I'm synchronizing my library with an Android phone and a Kodi box I'll add export and import to XSPF or at least some portable/relative path version of m3u. I'm not sure yet if I'll implement automatic playlist import/export since it may prove too much of a hassle (damn sync conflicts). The android app is just to accept additions to the playlists via the "share" interface and queue them to be added in the library by the quodlibet extension.

**Comment by declension** [comment]
See related #1122

**Comment by petko10** [comment]
I saw it, but I'm stuck on finding a very basic method - getting the songs library. In the existing plugins I see that they receive the selected songs (as a "songs" list), but I can't find an example with access to the whole library.

**Comment by petko10** [comment]
Figured it out, there's access to the app object when the "from quodlibet import app" . Attributes can be stored via the library object via altering _contents and save()-ing.
